scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,175 @@
+            if (config.runner.fatalWarnings) {
+              throw e
+            } else if (config.runner.ignoreWarnings) {
+              // do nothing
+              None
+            } else {
+              Some(onError(file, e))
+            }
+          case Formatted.Success(code) =>
+            Some(onFormat(file, input, code))
+        }
       }
+    )
   }
-  override def globalSettings: Seq[Def.Setting[_]] =
-    Seq(
-      commands += autoImport.scalafmt
-    ) ++
-      addCommandAlias("scalafmtTest", "scalafmt --test") ++
-      addCommandAlias("scalafmtDiffTest", "scalafmt --diff --test") ++
-      addCommandAlias("scalafmtDiff", "scalafmt --diff")
 
+  private def formatSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Unit = {
+    withFormattedSources(sources, config)(
+      (file, e) => log.error(s"Error in ${file.toString}: $e"),
+      (file, input, output) => {
+        if (input != output) {
+          log.info(s"Successfully formatted ${file.toString}.")
+          IO.write(file, output)
+        }
+      }
+    )
+    PlatformTokenizerCache.megaCache.clear()
+  }
+
+  private def checkSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Boolean = {
+    val res = withFormattedSources(sources, config)(
+      (file, e) => {
+        log.error(s"Error in ${file.toString}: $e")
+        false
+      },
+      (file, input, output) => {
+        val diff = input != output
+        if (diff) {
+          throw new MessageOnlyException(
+            s"${file.toString} isn't formatted properly!")
+        }
+        diff
+      }
+    ).flatten.forall(x => x)
+    PlatformTokenizerCache.megaCache.clear()
+    res
+  }
+
+  private lazy val sbtSources = thisProject.map(
+    proj => {
+      val rootSbt =
+        BuildPaths.configurationSources(proj.base).filterNot(_.isHidden)
+      val projectSbt =
+        (BuildPaths.projectStandard(proj.base) * GlobFilter("*.sbt")).get
+          .filterNot(_.isHidden)
+      rootSbt ++ projectSbt
+    }
+  )
+  private lazy val projectSources = thisProject.map(proj =>
+    (BuildPaths.projectStandard(proj.base) * GlobFilter("*.scala")).get)
+
+  lazy val scalafmtConfigSettings: Seq[Def.Setting[_]] = Seq(
+    scalafmt := formatSources(
+      (unmanagedSources in scalafmt).value,
+      scalaConfig.value,
+      streams.value.log),
+    scalafmtSbt := {
+      formatSources(sbtSources.value, sbtConfig.value, streams.value.log)
+      formatSources(projectSources.value, scalaConfig.value, streams.value.log)
+    },
+    scalafmtCheck :=
+      checkSources(
+        (unmanagedSources in scalafmt).value,
+        scalaConfig.value,
+        streams.value.log),
+    scalafmtSbtCheck := {
+      checkSources(sbtSources.value, sbtConfig.value, streams.value.log)
+      checkSources(projectSources.value, scalaConfig.value, streams.value.log)
+    },
+    compileInputs in compile := Def.taskDyn {
```

Comment:
iiuc, this implementation still breaks `inspect tree compile`, we need to push `Def.taskDyn` into a `scalafmtDoAutoformat` task, like is done here:

https://github.com/sbt/sbt-scalariform/blob/master/src/main/scala/com/typesafe/sbt/SbtScalariform.scala#L82-L87

I may be wrong @dwijnand may know better

Reply:
well, I just copied it from `neo-scalafmt-sbt` plugin :D

Reply:
Yes, `Def.taskDyn` "breaks" `inspect` - as in the dependencies don't show up.

Reply:
In addition, note that this could/should be a `settingDyn`, as the test only depends on the value of a setting--the fact that the returned thing is a task is irrelevant. (Too many tasks out there are `taskDyn`s where they could be `settingDyn`s.) If https://github.com/sbt/sbt/issues/3258 was implemented, this problem with breaking `inspect` wouldn't happen.